src/engine/email.go

```diff
@@ -0,0 +1,118 @@
+package engine
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+
+	"github.com/scoville/scvl/src/domain"
+)
+
+// CreateEmailRequest is the request for creating an email
+type CreateEmailRequest struct {
+	IsPreview      bool
+	SpreadsheetURL string
+	SheetName      string
+	Sender         string
+	Template       string
+	Title          string
+	User           *domain.User
+}
+
+// SendEmailRequest is the request for sending an email
+type SendEmailRequest struct {
+	ToAddresses  []*string
+	BccAddresses []*string
+	Body         string
+}
+
+// CreateEmail creates an email
+func (e *Engine) CreateEmail(req *CreateEmailRequest) (emailTemplate *domain.EmailTemplate, err error) {
+
+	emailTemplate, err = createEmailTemplate(e, req)
+	return
+}
+
+// SendEmail send a email
+func (e *Engine) SendEmail(req *CreateEmailRequest) (err error) {
+	emailTemplate, err := createEmailTemplate(e, req)
+	if err != nil {
+		return
+	}
+	// base64でエンコード済みの透過gifを本文に追加
+	emailID := strconv.Itoa(int(emailTemplate.ID))
+	openConfirmationCode := "<img src=\"https://scvl.site/emails/read/?id=" + emailID + "\">"
+	req.Template = req.Template + openConfirmationCode
+	err = e.awsClient.SendGroupEmails(emailTemplate.BatchEmail.Emails, emailTemplate.BatchEmail.Sender)
+	if err != nil {
+		return
+	}
+	for _, email := range emailTemplate.BatchEmail.Emails {
+		err = e.sqlClient.CreateEmail(email)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+// ReadEmail update emails opened_at
+func (e *Engine) ReadEmail(emailID string) (err error) {
+	err = e.sqlClient.ReadEmail(emailID)
+	return
+}
+
+func createEmailTemplate(e *Engine, req *CreateEmailRequest) (emailTemplate *domain.EmailTemplate, err error) {
```

Comment:
I think it's better to be implemented as `func (e *Engine) createEmailTemplate(req *CreateEmailRequest)`